### PR TITLE
Fix warnings produced when running test_optim.py

### LINKED ISSUE #67696: Warnings being produced when testing optim.
## 🐛 Bug

This is a small issue: pytorch tests have warnings when run on master (hash: 82f7f8d471241b0f417f932c59558774cc273041).

While this is clearly not a glaring issue, I think that tests ought not to produce warnings when running. It makes it harder for developers to see whether or not that is produced by their work.
Secondly, I think it is generally worth following the paradigms we expect to code with when we write tests.
Does this seem reasonable?

## To Reproduce

1. Install torch.
2. `cd pytorch`
3. `python test/test_optim.py`

Error messages:
```
/home/oliver/code/playground/pytorch-master/torch/optim/asgd.py:34: UserWarning: optimizer contains a parameter group with duplicate parameters; in future, this will cause an error; see github.com/pytorch/pytorch/issues/40967 for more information
  super(ASGD, self).__init__(params, defaults)
 ```
 This occurs for, as far as I can tell, every optimiser. As the error is the same, I will include only one copy.
 
 ```
 ........................../home/oliver/code/playground/pytorch-master/torch/optim/lr_scheduler.py:129: UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
  warnings.warn("Detected call of `lr_scheduler.step()` before `optimizer.step()`. "
```
Is produced several times too. Local testing makes me 99% sure this is caused in this snippet:
```
for _i in range(200):
            for scheduler in schedulers:
                if isinstance(scheduler, ReduceLROnPlateau):
                    val_loss = fn()
                    scheduler.step(val_loss)
                else:
                    scheduler.step()
            optimizer.step(fn)
```
[Source of this test code.](https://github.com/pytorch/pytorch/blob/82f7f8d471241b0f417f932c59558774cc273041/test/test_optim.py#L128)
  

## Expected behavior

Test should pass and issue no warnings.

## Environment

```
Collecting environment information...
PyTorch version: 1.11.0a0+gitcd51d2a
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~18.04) 9.4.0
Clang version: Could not collect # manual note, I built torch with `clang-6.0`
CMake version: version 3.19.6
Libc version: glibc-2.27

Python version: 3.8.12 (default, Oct 12 2021, 13:49:34)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-87-generic-x86_64-with-glibc2.17
Is CUDA available: False
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti
GPU 2: GeForce GTX 1080 Ti
GPU 3: GeForce GTX 1080 Ti
GPU 4: GeForce GTX 1080 Ti
GPU 5: GeForce GTX 1080 Ti
GPU 6: GeForce GTX 1080 Ti
GPU 7: GeForce GTX 1080 Ti

Nvidia driver version: 450.102.04
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.21.2
[pip3] torch==1.11.0a0+git82f7f8d
[conda] blas                      1.0                         mkl
[conda] mkl                       2021.4.0           h06a4308_640
[conda] mkl-include               2021.4.0           h06a4308_640
[conda] mkl-service               2.4.0            py38h7f8727e_0
[conda] mkl_fft                   1.3.1            py38hd3c417c_0
[conda] mkl_random                1.2.2            py38h51133e4_0
[conda] numpy                     1.21.2           py38h20f2e39_0
[conda] numpy-base                1.21.2           py38h79a1101_0
[conda] torch                     1.11.0a0+git82f7f8d           dev_0    <develop>
```

## Additional context

One of the two issues is linked to #40967 -- a fix was made to emit a warning, but our tests were not updated.
I am happy to fix all of these issues, if that would be helpful!

cc @vincentqb @jbschlosser @albanD @mruberry

### PR DESCRIPTION
**Pull request by OliverFM**
Fixes part of #67696 by adding calls to `optimizer.step()` in various places.

## Notes for reviewers:
- It is not entirely clear which is the right optimizer to step in each case. I have favoured the more explicit approach of creating a set of optimizers and calling step on each of them.
- At the time of writing, the only Scheduler without an `optimizer` instance variable is `ChainedScheduler` which I need to deal with once. I use `hasattr` to do this check. Let me know if this ought to be changed.
- I am opening this PR for review when it only solve part of the issue, as I'd rather get feedback sooner. I think it is fine to fix the issue in several PRs too.

### CODE REVIEW

**Review by albanD — APPROVED** [review]
Looks good. Thanks for fixing all of these!

**Comment by OliverFM** [comment]
> Looks good. Thanks for fixing all of these!

Great! Do I need to do anything further to get this merged in, @albanD?

**Comment by albanD** [comment]
Nope, this is on my end to merge it. Which just happened, so the bot will close this very soon.

**Comment by OliverFM** [comment]
@albanD I was working on adding `maximize` to adam, and it looks like this change got reverted somewhere down the line -- since this warnings are being produced again. Any idea why that would be? and if it can be reapplied?

https://github.com/pytorch/pytorch/blob/a6a2616558e996e6953ffcb246e98ce31a7c570d/test/test_optim.py#L139

**Comment by albanD** [comment]
Hmmm not sure. Did the PR adding maximize reverted this by mistake.

**Comment by OliverFM** [comment]
> Hmmm not sure. Did the PR adding maximize reverted this by mistake?

Definitely possible. I can make a new PR to reapply the changes.
